components/raftstore/src/store/peer.rs

```diff
@@ -3391,13 +3382,41 @@ where
             return;
         }
 
-        // Reverse sort peers based on `next_idx`, then try to get a potential quorum.
-        next_idxs.sort_by(|x, y| y.1.cmp(&x.1));
-        let (mut potential_quorum, mut quorum_ok) = (HashSet::default(), false);
-        for &(peer_id, _, usage) in &next_idxs {
-            if matches!(usage, Some(DiskUsage::AlreadyFull)) {
-                continue;
+        // Reverse sort peers based on `next_idx` and `usage`, then try to get a potential quorum.
+        next_idxs.sort_by(|x, y| {
+            if y.2.is_some() {
+                if matches!(y.2.unwrap(), DiskUsage::AlreadyFull) {
+                    if x.2.is_some() {
+                        if matches!(x.2.unwrap(), DiskUsage::AlreadyFull) {
+                            return y.1.cmp(&x.1);
+                        } else {
+                            return std::cmp::Ordering::Less;
+                        }
+                    } else {
+                        return std::cmp::Ordering::Less;
+                    }
+                } else {
+                    if x.2.is_some() {
+                        if matches!(x.2.unwrap(), DiskUsage::AlreadyFull) {
+                            return std::cmp::Ordering::Greater;
+                        } else {
+                            return y.1.cmp(&x.1);
+                        }
+                    } else {
+                        return std::cmp::Ordering::Less;
+                    }
+                }
+            } else {
+                if x.2.is_some() {
+                    return std::cmp::Ordering::Greater;
+                } else {
+                    return y.1.cmp(&x.1);
+                }
             }
+        });
+
+        let (mut potential_quorum, mut quorum_ok) = (HashSet::default(), false);
+        for &(peer_id, _, _) in &next_idxs {
```

Comment:
It's also necessary to add a new field `emergency` into `DiskFullPeers`. If the field is true it means some peers in the potential quorum are almost disk full, so that only conf changes are allowed.

Reply:
And don't forget to add a test to show that only configuration changes and admin requests are allowed if the potential quorum contains any disk almost full peers.